scantool/scantool.c

```diff
@@ -640,6 +640,15 @@ l2_do_send(struct diag_l2_conn *d_conn, void *data, size_t len, void *handle)
 
 	msg.len = len;
 	msg.data = (uint8_t *)data;
+
+	if (d_conn->l2proto->diag_l2_protocol == DIAG_L2_PROT_VAG) {
```

Comment:
Consider an example request message for J1979 over KWP2000. The tester sends Request Current Powertrain Diagnostic Data for PID $00. The message (including checksum) is 82 33 F1 01 00 A7. l2_iso14230 considers 82 33 F1 to be the header and 01 00 to be the data. The Service ID ($01 for Request Current Powertrain Diagnostic Data) is the first data byte.

On the other hand, consider an example request message for KWP71. The tester sends Read ADC for ADC channel 2. Assume the message sequence number is 1. The message (including EOT) is 04 01 08 02 03. l2_vag considers 04 01 08 to be the header and 02 to be the data. The Block Title ($08 for Read ADC) is the type field of the header.

That example request message will work on a Motronic ECU from a Volvo 850 with KWP71, but it is also valid for a Volkswagen with KWP1281, or a BMW with KWP71, etc. So it applies to diag sendreq in addition to 850 sendreq.

For sendreq, msg is created in a local variable in l2_do_send. The only way for sendreq to control msg.type is either to put the value in data[] or to add a parameter to diag_l2_send for msg.type.

Maybe l2_vag should be changed to be consistent with the other L2s, but that's a separate issue.